shims/spark32/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java

```diff
@@ -0,0 +1,928 @@
+      // this is necessary.
+      return 0L;
+    }
+
+    logger.info(
+        "Thread {} force spilling sort data of {} to disk ({} {} so far)",
+        Thread.currentThread().getId(),
+        Utils.bytesToString(getMemoryUsage()),
+        spillWriters.size(),
+        spillWriters.size() > 1 ? " times" : " time");
+
+    ShuffleWriteMetrics writeMetrics = new ShuffleWriteMetrics();
+
+    final UnsafeSorterSpillWriter spillWriter =
+        new UnsafeSorterSpillWriter(
+            blockManager, fileBufferSizeBytes, writeMetrics, inMemSorter.numRecords());
+    spillWriters.add(spillWriter);
+    spillIterator(inMemSorter.getSortedIterator(), spillWriter);
+
+    final long spillSize = freeMemory();
+    // Note that this is more-or-less going to be a multiple of the page size, so wasted space in
+    // pages will currently be counted as memory spilled even though that space isn't actually
+    // written to disk. This also counts the space needed to store the sorter's pointer array.
+    inMemSorter.freeMemory();
+    // Reset the in-memory sorter's pointer array only after freeing up the memory pages holding the
+    // records. Otherwise, if the task is over allocated memory, then without freeing the memory
+    // pages, we might not be able to get memory for the pointer array.
+
+    taskContext.taskMetrics().incMemoryBytesSpilled(spillSize);
+    taskContext.taskMetrics().incDiskBytesSpilled(writeMetrics.bytesWritten());
+    totalSpillBytes += spillSize;
+    return spillSize;
+  }
+
+  /** Sort and spill the current records in response to memory pressure. */
+  @Override
+  public long spill(long size, MemoryConsumer trigger) throws IOException {
+    if (trigger != this) {
+      if (readingIterator != null) {
+        return readingIterator.spill();
+      }
+      return 0L; // this should throw exception
+    }
+
+    if (inMemSorter == null || inMemSorter.numRecords() <= 0) {
+      // There could still be some memory allocated when there are no records in the in-memory
+      // sorter. We will not spill it however, to ensure that we can always process at least one
+      // record before spilling. See the comments in `allocateMemoryForRecordIfNecessary` for why
+      // this is necessary.
+      return 0L;
+    }
+
+    logger.info(
+        "Thread {} spilling sort data of {} to disk ({} {} so far)",
+        Thread.currentThread().getId(),
+        Utils.bytesToString(getMemoryUsage()),
+        spillWriters.size(),
+        spillWriters.size() > 1 ? " times" : " time");
+
+    ShuffleWriteMetrics writeMetrics = new ShuffleWriteMetrics();
+
+    final UnsafeSorterSpillWriter spillWriter =
+        new UnsafeSorterSpillWriter(
+            blockManager, fileBufferSizeBytes, writeMetrics, inMemSorter.numRecords());
+    logger.warn("inMemSorter numRecords: " + inMemSorter.numRecords());
+    spillWriters.add(spillWriter);
+    UnsafeSorterIterator iterator = inMemSorter.getSortedIterator();
```

Comment:
@jinchengchenghh The error is caused by the mismatch of inMemSorter.numRecords() vs. iterator.getNumRecords(). What's the reason?

```
24/08/03 05:22:17 INFO [Thread-14] sort.UnsafeExternalSorter: Thread 83 force spilling sort data of 12.3 GiB to disk (1  time so far)
24/08/03 05:22:17 WARN [Thread-14] sort.UnsafeExternalSorter: UnsafeSorterSpillWriter numRecordsToWrite : 48823146
24/08/03 05:23:38 WARN [Executor task launch worker for task 421.0 in stage 0.0 (TID 421)] sort.UnsafeExternalSorter: inMemSorter numRecords: 44739242
24/08/03 05:23:38 WARN [Executor task launch worker for task 421.0 in stage 0.0 (TID 421)] sort.UnsafeExternalSorter: iterator numRecords: 44739242
24/08/03 05:23:38 WARN [Executor task launch worker for task 421.0 in stage 0.0 (TID 421)] sort.UnsafeExternalSorter: inMemSorter SortedIterator length 44739242
24/08/03 05:23:38 WARN [Executor task launch worker for task 421.0 in stage 0.0 (TID 421)] sort.UnsafeExternalSorter: inMemIterator size: 44739242
24/08/03 05:23:40 WARN [Executor task launch worker for task 423.0 in stage 0.0 (TID 423)] sort.UnsafeExternalSorter: inMemSorter numRecords: 44739242
24/08/03 05:23:40 WARN [Executor task launch worker for task 423.0 in stage 0.0 (TID 423)] sort.UnsafeExternalSorter: iterator numRecords: 44739242
24/08/03 05:23:40 WARN [Executor task launch worker for task 423.0 in stage 0.0 (TID 423)] sort.UnsafeExternalSorter: inMemSorter SortedIterator length 44739242
24/08/03 05:23:40 WARN [Executor task launch worker for task 423.0 in stage 0.0 (TID 423)] sort.UnsafeExternalSorter: inMemIterator size: 44739242
24/08/03 05:23:40 WARN [Executor task launch worker for task 420.0 in stage 0.0 (TID 420)] sort.UnsafeExternalSorter: inMemSorter numRecords: 44739242
24/08/03 05:23:40 WARN [Executor task launch worker for task 420.0 in stage 0.0 (TID 420)] sort.UnsafeExternalSorter: iterator numRecords: 44739242
24/08/03 05:23:40 WARN [Executor task launch worker for task 420.0 in stage 0.0 (TID 420)] sort.UnsafeExternalSorter: inMemSorter SortedIterator length 44739242
24/08/03 05:23:40 WARN [Executor task launch worker for task 420.0 in stage 0.0 (TID 420)] sort.UnsafeExternalSorter: inMemIterator size: 44739242
24/08/03 05:23:47 WARN [Thread-14] sort.UnsafeExternalSorter: inMemIterator size: 49009940
24/08/03 05:24:09 ERROR [Thread-14] listener.ManagedReservationListener: Error reserving memory from target
java.lang.IllegalStateException: Number of records written exceeded numRecordsToWrite = 48823146
	at org.apache.spark.util.collection.unsafe.sort.UnsafeSorterSpillWriter.write(UnsafeSorterSpillWriter.java:118)
	at org.apache.spark.util.collection.unsafe.sort.UnsafeExternalSorter.spillIterator(UnsafeExternalSorter.java:643)
```